internal/jimm/model_cleanup_test.go

```diff
@@ -0,0 +1,195 @@
+  life: alive
+  users:
+  - user: alice@canonical.com
+    access: admin
+  - user: bob@canonical.com
+    access: write
+users:
+- username: alice@canonical.com
+  controller-access: superuser
+`
+
+func TestPollModelsDying(t *testing.T) {
+	// init
+	c := qt.New(t)
+	ctx := context.Background()
+
+	client, _, _, err := jimmtest.SetupTestOFGAClient(c.Name())
+	c.Assert(err, qt.IsNil)
+	j := &jimm.JIMM{
+		UUID:          uuid.NewString(),
+		OpenFGAClient: client,
+		Database: db.Database{
+			DB: jimmtest.PostgresDB(c, nil),
+		},
+	}
+	err = j.Database.Migrate(ctx, false)
+	c.Assert(err, qt.IsNil)
+	jimmAdmin, err := j.GetUser(ctx, "alice@canonical.com")
+	c.Assert(err, qt.IsNil)
+
+	env := jimmtest.ParseEnvironment(c, modelPollerTestEnv)
+	env.PopulateDBAndPermissions(c, j.ResourceTag(), j.Database, client)
+
+	j.Dialer = &jimmtest.Dialer{
+		API: &jimmtest.API{
+			ModelInfo_: func(ctx context.Context, mi *jujuparams.ModelInfo) error {
+				switch mi.UUID {
+				case env.Models[0].UUID:
+					return errors.E(errors.CodeNotFound)
+				case env.Models[1].UUID:
+					return nil
+				default:
+					return errors.E("new error")
+				}
+			},
+			DestroyModel_: func(ctx context.Context, mt names.ModelTag, b1, b2 *bool, d1, d2 *time.Duration) error {
+				return nil
+			},
+		},
+	}
+	err = j.DestroyModel(ctx, jimmAdmin, names.NewModelTag(env.Models[0].UUID), nil, nil, nil, nil)
+	c.Assert(err, qt.IsNil)
+
+	// test
+	err = j.CleanupDyingModels(ctx)
+	c.Assert(err, qt.IsNil)
+
+	model := dbmodel.Model{
+		UUID: sql.NullString{
+			String: env.Models[0].UUID,
+			Valid:  true,
+		},
+	}
+	err = j.DB().GetModel(ctx, &model)
+	c.Assert(err, qt.ErrorMatches, "model not found")
+
+	model = dbmodel.Model{
+		UUID: sql.NullString{
+			String: env.Models[1].UUID,
+			Valid:  true,
+		},
+	}
+	err = j.DB().GetModel(ctx, &model)
+	c.Assert(err, qt.IsNil)
+}
+
+func TestPollModelsDyingControllerErrors(t *testing.T) {
+	// init
+	c := qt.New(t)
+	ctx := context.Background()
+
+	client, _, _, err := jimmtest.SetupTestOFGAClient(c.Name())
+	c.Assert(err, qt.IsNil)
+	j := &jimm.JIMM{
+		UUID:          uuid.NewString(),
+		OpenFGAClient: client,
+		Database: db.Database{
+			DB: jimmtest.PostgresDB(c, nil),
+		},
+	}
+	err = j.Database.Migrate(ctx, false)
+	c.Assert(err, qt.IsNil)
+
+	env := jimmtest.ParseEnvironment(c, modelPollerTestEnv)
+	env.PopulateDBAndPermissions(c, j.ResourceTag(), j.Database, client)
+	jimmAdmin, err := j.GetUser(ctx, "alice@canonical.com")
+	c.Assert(err, qt.IsNil)
+	j.Dialer = &jimmtest.Dialer{
+		API: &jimmtest.API{
+			ModelInfo_: func(ctx context.Context, mi *jujuparams.ModelInfo) error {
+				return errors.E("controller not available")
+			},
+			DestroyModel_: func(ctx context.Context, mt names.ModelTag, b1, b2 *bool, d1, d2 *time.Duration) error {
+				return nil
+			},
+		},
+	}
+	err = j.DestroyModel(ctx, jimmAdmin, names.NewModelTag(env.Models[0].UUID), nil, nil, nil, nil)
+	c.Assert(err, qt.IsNil)
+
+	// test
```

Comment:
Can just put the rest in a setup function and remove this comment? Yeah actually, make a suite please it's much cleaner. See role for example.

Reply:
done it